File: binary_search_tree/swift/BinarySearchTree.playground/Sources/BinarySearchTree.swift
import Foundation

public class Node<T: Equatable & Comparable> {
  fileprivate var value: T
  fileprivate private(set) var count: Int
  fileprivate var left: Node<T>?
  fileprivate var right: Node<T>?

  fileprivate init(_ value: T) {
    self.value = value
    self.count = 1
  }

  func search(for value: T) -> Node<T>? {
    if (self.value == value) {
      return self
    } else if
        value < self.value,
        let left = self.left
    {
        return left.search(for: value)
    } else if let right = self.right {
        return right.search(for: value)
    } else {
        return nil
    }
  }

  func addLeftChild(_ value: T) {
    if let left = self.left {
      left.addChild(value)
    } else {
      self.left = Node(value)
    }
  }

  func addRightChild(_ value: T) {
    if let right = self.right {
      right.addChild(value)
    } else {
      self.right = Node(value)
    }
  }

  fileprivate func addChild(_ value: T) {
    if value == self.value {
      self.count += 1
    } else if (value < self.value) {
      self.addLeftChild(value)
    } else {
      self.addRightChild(value)
    }
  }
}

public class BinarySearchTree<T: Equatable & Comparable> {
  private var root: Node<T>?

  public init() {}

  public func addNode(withValue value: T) {
    if let root = self.root {
      root.addChild(value)
    } else {
      self.root = Node(value)
    }
  }

  public func search(for value: T) -> Node<T>? {
    return self.root?.search(for: value)
  }
}

File: binary_search_tree/binary_search_tree.js
class Node {
  constructor(value) {
    this.value = value;
    this.count = 1;
    this.left = null;
    this.right = null;
  }

  search = value => {
    if (value == this.value) {
      return this;
    } else if (value < this.value && this.left) {
      return this.left.search(value);
    } else if (this.right) {
      return this.right.search(value);
    } else {
      return null;
    }
  };

  addLeftChild = child => {
    if (this.left) {
      this.left.addChild(child);
    } else {
      this.left = child;
    }
  };

  addRightChild = child => {
    if (this.right) {
      this.right.addChild(child);
    } else {
      this.right = child;
    }
  };

  addChild = child => {
    if (child.value == this.value) {
      this.count++;
    } else if (child.value < this.value) {
      addLeftChild(child);
    } else {
      addRightChild(child);
    }
  };
}

class Tree {
  constructor() {
    this.root = null;
  }

  addNode = value => {
    let newNode = new Node(value);

    if (this.root) {
      this.root.addChild(newNode);
    } else {
      this.root = newNode;
    }
  };

  search = value => {
    return this.root.search(value);
  };
}
